File: tools/tf-types-ts/src/core/revocation.ts
import type { Revocation } from "../generated/revocation.js";

export type RevocationTargetKind = Revocation["target_kind"];

export interface RevocationQuery {
  readonly id: string;
  readonly kind: RevocationTargetKind;
}

export class RevocationIndex {
  private readonly byKind: Map<RevocationTargetKind, Map<string, Revocation>>;

  private constructor(byKind: Map<RevocationTargetKind, Map<string, Revocation>>) {
    this.byKind = byKind;
  }

  static from(revocations: readonly Revocation[]): RevocationIndex {
    const byKind = new Map<RevocationTargetKind, Map<string, Revocation>>();
    for (const r of revocations) {
      let bucket = byKind.get(r.target_kind);
      if (!bucket) {
        bucket = new Map();
        byKind.set(r.target_kind, bucket);
      }
      const existing = bucket.get(r.target_id);
      if (!existing || existing.effective_at > r.effective_at) {
        bucket.set(r.target_id, r);
      }
    }
    return new RevocationIndex(byKind);
  }

  isRevoked(target: RevocationQuery, at: string): boolean {
    const bucket = this.byKind.get(target.kind);
    if (!bucket) return false;
    const r = bucket.get(target.id);
    if (!r) return false;
    return r.effective_at <= at;
  }
}

File: crates/tf-types/src/revocation.rs
//! Revocation index — mirrors `tools/tf-types-ts/src/core/revocation.ts`.

use std::collections::HashMap;

use crate::generated::revocation::{Revocation, Revocation_TargetKind};

pub struct RevocationIndex {
    by_kind: HashMap<Revocation_TargetKind, HashMap<String, Revocation>>,
}

impl RevocationIndex {
    pub fn from_slice(revocations: &[Revocation]) -> Self {
        let mut by_kind: HashMap<Revocation_TargetKind, HashMap<String, Revocation>> =
            HashMap::new();
        for r in revocations {
            let bucket = by_kind.entry(r.target_kind.clone()).or_default();
            match bucket.get(&r.target_id) {
                Some(existing) if existing.effective_at <= r.effective_at => {}
                _ => {
                    bucket.insert(r.target_id.clone(), r.clone());
                }
            }
        }
        RevocationIndex { by_kind }
    }

    pub fn is_revoked(&self, id: &str, kind: &Revocation_TargetKind, at: &str) -> bool {
        let Some(bucket) = self.by_kind.get(kind) else {
            return false;
        };
        let Some(rev) = bucket.get(id) else {
            return false;
        };
        rev.effective_at.as_str() <= at
    }
}

impl std::hash::Hash for Revocation_TargetKind {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        std::mem::discriminant(self).hash(state);
    }
}
